vars/runPCT.groovy

```diff
@@ -0,0 +1,131 @@
+                metadata = [:]
+            }
+            // Allow to override pctUrl and pctRevision from metadata file
+            pctUrl = metadata.pctUrl ?: pctUrl
+            pctRevision = metadata.pctRevision ?: pctRevision
+            isPublishedImage = pctUrl.startsWith("docker://")
+            isLocalPCT = pctUrl.startsWith("file://")
+
+            // Allow override of jenkins version from metadata file
+            jenkins = metadata.jenkins ?: jenkins
+
+            if (!(isPublishedImage || isLocalPCT)) {
+                echo 'Checking connectivity to PCT sources…'
+                sh "git ls-remote --exit-code -h ${pctUrl}"
+            }
+            // Validation ended
+
+            // Jenkins war
+            infra.stashJenkinsWar(jenkins)
+            // PCT
+            // Only validate PCT if not running a remote docker image
+            if (!isPublishedImage) {
+                if (isLocalPCT) { // Deal with already existing PCT sources
+                    pctSourcesFolder = pctUrl - "file://"
+                } else {
+                    dir(pctSourcesFolder) {
+                        checkout changelog: true, poll: false, scm: [$class           : 'GitSCM', branches: [[name: pctRevision]],
+                                                                     userRemoteConfigs: [[url: pctUrl]]]
+                    }
+                }
+                dir(pctSourcesFolder) {
+                    // We may need to run things in parallel later, so avoid several checkouts
+                    stash name: "pct"
+                }
+            }
+        }
+    })
+
+    if (skipExecution) {
+        return
+    }
+
+    infra.ensureInNode(env, env.RUN_PCT_DOCKER_NODE ?: 'docker', {
+
+        stage("Running PCT") {
+            if (!isPublishedImage) {
+                dir("pctSources") {
+                    unstash "pct"
+                    pctContainerImage = docker.build("jenkins/pct:${env.BUILD_NUMBER}")
+                }
+            } else {
+                def tag = pctUrl - "docker://"
+                pctContainerImage = docker.image(tag)
+            }
+
+            def plugins = metadata.plugins
+
+            def localSnapshots = metadata.useLocalSnapshots != null ? metadata.useLocalSnapshots : true
+
+            def testingBranches = [:]
+            def containerArgsBase = "-v /var/run/docker.sock:/var/run/docker.sock -v jenkins.war:/pct/jenkins.war:ro -u root"
+            for (def i = 0; i < plugins.size(); i++) {
+                def plugin = plugins[i]
+                testingBranches["PCT-${plugin}"] = {
+                    unstash "jenkinsWar"
+                    pctContainerImage.inside(containerArgsBase) {
+                        def command = 'run-pct'
+                        if (localSnapshots && localPluginsStashName) {
+                            dir("localPlugins") {
+                                unstash name: localPluginsStashName
```

Comment:
`localSnapshots` is `true` by default, but this method does not initialize local Plugin Stash itself. It means that this operation is guaranteed to fail if you invoke the methods with default arguments without externally-defined stashes.

I would change the default value of `localSnapshots ` to `false`. 🐛

Reply:
Actually `runATH()` has the same behavior. Do not consider it as a blocker, but it was a bit unexpected

Reply:
After thinking a little more about this I am going to stay with the original defaults, the fact that is going to fail if you do not set `useLocalSnapshots` to false and you do not stash the plugins is already documented since the beginning and as I said I believe that default value makes much more sense

Reply:
Agreed. It's documented now, so should work